sys/net/network_layer/ng_sixlowpan/frag/rbuf.c

```diff
@@ -0,0 +1,307 @@
+    bool first;
+
+    switch (frag->disp_size.u8[0] & NG_SIXLOWPAN_FRAG_DISP_MASK) {
+        case NG_SIXLOWPAN_FRAG_1_DISP:
+            tag = byteorder_ntohs(frag->tag);
+            datagram_size = byteorder_ntohs(frag->disp_size) & NG_SIXLOWPAN_FRAG_SIZE_MASK;
+            frag_size = (pkt->size - sizeof(ng_sixlowpan_frag_t));
+            data = (uint8_t *)(hdr + 1);
+            first = true;
+            break;
+
+        case NG_SIXLOWPAN_FRAG_N_DISP:
+            tag = byteorder_ntohs(frag->tag);
+            datagram_size = byteorder_ntohs(frag->disp_size) & NG_SIXLOWPAN_FRAG_SIZE_MASK;
+            offset = (((ng_sixlowpan_frag_n_t *)frag)->offset * 8);
+            frag_size = (pkt->size - sizeof(ng_sixlowpan_frag_n_t));
+            data = (uint8_t *)(hdr + 1) + 1;
+            break;
+
+        default:
+            DEBUG("6lo rbuf: Not a fragment header.\n");
+            ng_pktbuf_release(pkt);
+
+            return;
+    }
+
+    _rbuf_gc();
+    entry = _rbuf_get(ng_netif_hdr_get_src_addr(hdr), hdr->src_l2addr_len,
+                      ng_netif_hdr_get_dst_addr(hdr), hdr->dst_l2addr_len,
+                      datagram_size, tag);
+
+    if (entry == NULL) {
+        DEBUG("6lo rbuf: reassembly buffer full.\n");
+        ng_pktbuf_release(pkt);
+
+        return;
+    }
+
+    if (_rbuf_update_ints(entry, offset, frag_size)) {
+        if (first) {
+            /* some dispatches do not count to datagram size and we need
+             * more space because of that */
+            switch (data[0]) {
+                case NG_SIXLOWPAN_UNCOMPRESSED:
+                    if (ng_pktbuf_realloc_data(pkt, pkt->size + 1) < 0) {
+                        DEBUG("6lo rbuf: could not reallocate packet data.\n");
+                        ng_pktbuf_release(pkt);
+                        return;
+                    }
+
+                    /* move already inserted fragments 1 to the right */
+                    for (int i = pkt->size - 1; i > 0; i--) {
+                        uint8_t *d = ((uint8_t *)(pkt->data)) + i;
+                        *d = *(d - 1);
+                    }
+            }
+        }
+
+        DEBUG("6lo rbuf: add fragment data\n");
+        memcpy(((uint8_t *)entry->pkt->data) + offset,
+               (uint8_t *)(frag + 1), frag_size);
+        entry->cur_size += frag_size;
+    }
+
+    ng_pktbuf_release(pkt);
+
+    if (entry->cur_size == entry->pkt->size) {
+        ng_pktsnip_t *netif = ng_netif_hdr_build(entry->src, entry->src_len,
+                              entry->dst, entry->dst_len);
+
+        if (netif == NULL) {
+            DEBUG("6lo rbuf: error allocating netif header\n");
+            ng_pktbuf_release(entry->pkt);
+            return;
+        }
+
+        hdr = netif->data;
+        hdr->if_pid = iface;
+        entry->pkt->next = netif;
+
+        DEBUG("6lo rbuf: datagram complete, send to self\n");
+        ng_netapi_send(thread_getpid(), entry->pkt);
+        _rbuf_rem(entry);
```

Comment:
Since I still haven't understood 100% how this is working, I better ask: we don't need to release `pkt->entry` here?